misk/src/main/kotlin/misk/healthchecks/ClusterWideHealthService.kt

```diff
@@ -0,0 +1,114 @@
+import misk.healthchecks.ClusterWideHealthService.Companion.PERIOD
+import misk.healthchecks.ClusterWideHealthService.Companion.TOPIC
+import misk.logging.getLogger
+import misk.web.ConnectWebSocket
+import misk.web.actions.WebAction
+import misk.web.actions.WebSocket
+import misk.web.actions.WebSocketListener
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private val logger = getLogger<ClusterWideHealthService>()
+
+/**
+ * ClusterWideHealthService publishes its health information on [TOPIC] every
+ * [PERIOD]. If it does not receive a health status from a host within [PERIOD] * 2,
+ * it will remove it.
+ */
+@Singleton
+class ClusterWideHealthService : AbstractIdleService(), WebAction {
+  @Inject lateinit var kubernetesConfig: KubernetesConfig
+  @Inject lateinit var eventRouter: EventRouter
+  @Inject lateinit var services: List<Service>
+  @Inject lateinit var healthChecks: List<HealthCheck>
+  @Inject lateinit var moshi: Moshi
+  @Inject lateinit var gson: Gson
+  private val healthEventJsonAdapter get() = moshi.adapter(HealthEvent::class.java)
+  // TODO: concurrency?
+  private val subscribers = mutableListOf<WebSocket>()
+
+  @Inject
+  @ForClusterWideHealthService
+  lateinit var scheduledExecutorService: ScheduledExecutorService
+
+  val healthCheckCache: Cache<String, HealthEvent> = CacheBuilder.newBuilder()
+      .expireAfterWrite(PERIOD * 2, PERIOD_TIMEUNIT)
+      .removalListener<String, HealthEvent> { sendClusterHealth() }
+      .build<String, HealthEvent>()
+
+  @ConnectWebSocket("/health")
+  fun health(webSocket: WebSocket): WebSocketListener {
+    subscribers.add(webSocket)
+    sendClusterHealth()
+    return object : WebSocketListener() {}
+  }
+
+  override fun startUp() {
+    logger.info { "starting up ClusterWideHealthService" }
+    scheduledExecutorService.scheduleAtFixedRate({
+      val json = healthEventJsonAdapter.toJson(
+          HealthEvent(
+              kubernetesConfig.my_pod_name,
+              healthChecks.all { it.status().isHealthy }
+          )
+      )
+      eventRouter.getTopic<String>(TOPIC).publish(json)
+    }, 0, PERIOD, PERIOD_TIMEUNIT)
+    scheduledExecutorService.schedule({ subscribe() }, 3, TimeUnit.SECONDS)
+  }
+
+  override fun shutDown() {
+    logger.info { "shutting down ClusterWideHealthService" }
+    scheduledExecutorService.shutdown()
+  }
+
+  private fun sendClusterHealth() {
+    subscribers.forEach {
+      it.send(gson.toJson(healthCheckCache.asMap().toSortedMap()))
+    }
+  }
+
+  private fun subscribe() {
+    val topic = eventRouter.getTopic<String>(TOPIC)
+
+    val listener = object : Listener<String> {
+      override fun onEvent(subscription: Subscription<String>, event: String) {
+        val healthEvent = healthEventJsonAdapter.fromJson(event)!!
+        healthCheckCache.put(healthEvent.hostname, healthEvent)
+        sendClusterHealth()
```

Comment:
Should this method say who it’s sending it to? We can send it to the cluster, or to the connected websockets

Reply:
Maybe this is `clusterHealthChanged()` ?